scrapy/core/engine.py

```diff
@@ -181,14 +190,19 @@ def _next_request(self) -> None:
                 request = next(self.slot.start_requests)
             except StopIteration:
                 self.slot.start_requests = None
+                self.signals.send_catch_log(signal=signals.start_requests_exhausted)
             except Exception:
                 self.slot.start_requests = None
                 logger.error(
                     "Error while obtaining start requests",
                     exc_info=True,
                     extra={"spider": self.spider},
                 )
+                self.signals.send_catch_log(signal=signals.start_requests_exhausted)
```

Comment:
I wonder if we should implement separate method for these, but it felt a bit out of scope.